### ISSUE TITLE
[ENH]: Good looking date/time axes with primary and secondary annotation and interval annotation

### ISSUE THREAD

**Issue by eelcodoornbos** [matplotlib/matplotlib#27394]
### Problem

The default matplotlib date/time axes are difficult to customize to look professional. Date values usually denote time intervals (days, months, years), but matplotlib default date labels are attached to single ticks that mark the start of that interval. And since all date/time annotations are usually put on a single axis, the labels get long, and often have to be plotted at an angle, which does not look pleasing and makes the plots more difficult to read.

### Proposed solution

Good-looking and easy to read date/time axes need implementation of two features:
- Interval labels - where the labels for years, months, week numbers and days are drawn between two adjacent tick marks
- Primary and secondary axes that are vertically stacked, so that the top ticks/labels can denote a smaller unit of time than the bottom one.

Some excellent examples of the look that can be achieved using these two features can be seen in the cookbook of the Generic Mapping Tools: https://docs.generic-mapping-tools.org/latest/cookbook/options.html#cartesian-time-axes

The "interval labels" feature could also be made reusable for custom interval annotations, like in the further example on that website: https://docs.generic-mapping-tools.org/latest/cookbook/options.html#custom-axes

**Comment by rcomer** [comment]
Does `ConciseDateConverter` address your "primary and secondary axes" point?
https://matplotlib.org/stable/gallery/ticks/date_concise_formatter.html

**Comment by eelcodoornbos** [comment]
No, `ConciseDateConverter` does not add a secondary axis, only a single additional label. I find the examples not very intuitive to read, as this additional label is not a commonly used convention. The reader first has to parse what it means. For the example where the time series ends on April 1, 2005, it gets particularly confusing, since none of the dates displayed on the axis actually fall inside that month. 
The examples with the true primary and secondary axes in the Generic Mapping Tools documentation on the other hand, convey a much more conventional calendar-like approach to displaying dates and times, that should be clear to readers of the plots much more quickly.

**Comment by jklymak** [comment]
We do not currently have a way to add an extra level of ticks.  

You can fake this with a secondary x axis or with minor and major ticks on the main axis.  If I were doing it, I'd place a secondary xaxis at y=0 and just add carriage returns to the tick labels rather than messing with offsets by hand. 

Labelling time in general is quite hard and there will always be taste differences.  If your taste differs from the available defaults, you are going to have to do some work.   

The general request for multiple label levels is the same as https://github.com/matplotlib/matplotlib/issues/6321

**Comment by eelcodoornbos** [comment]
Thanks for those considerations. The [Stack Overflow example](https://stackoverflow.com/questions/19184484/how-to-add-group-labels-for-bar-charts) mentioned in #6321, already implements both dual labels and interval labelling with some additional code.
So then I think a good solution would be an alternative to [date_concise_formatter.html](https://matplotlib.org/stable/gallery/ticks/date_concise_formatter.html), which implements these features "behind-the-scenes" to offer two levels of annotation, and adds some sensible defaults based on the plot width and time interval for casual users, as well as offers parameters to customize the primary and secondary date/time annotations.

I agree that labelling time is hard, and the above would take a considerable implementation and fine-tuning effort. But I don't think the main differences between the examples under the links I gave above on the one hand, and those in [date_concise_formatter.html](https://matplotlib.org/stable/gallery/ticks/date_concise_formatter.html) and the current matplotlib default on the other, come down purely to a matter of taste. I think the readability/usability differences can be considered objectively.

**Comment by eelcodoornbos** [comment]
Yes, something like that would be a very good starting point. 

The default locators/formatters for https://spaceweather.knmi.nl/viewer/ are implemented (in JS) here: https://gitlab.com/KNMI-OSS/spaceweather/knmi-hapi-timeline-viewer/-/blob/main/src/lib/svg_plots/plot_frame_components/datetimetickinfo.js
These were created by trial/error/tweaking, but it seems to scale well across a large range of plot canvas sizes, from smartphones to large high-res displays. Something like that could be translated to use the matplotlib locators/formatters, and would probably have to be tweaked a bit further to correct for font/plot size/unit differences between this JS/CSS/SVG implementation and matplotlib. Those could then feed into the DateCompositeLocator/Formatters.

**Comment by jklymak** [comment]
The aesthetics of the date locators and formatters is a side issue to me - anyone can write their own to override the Matplotlib-offered ones, and the defaults aren't going to change until we do a style-break release.

The first job is deciding on an API for how to add a second (or more) level of tick annotations.  

> `DateCompositeLocator([WeekdayLocator(**params),  MicrosecondLocator(params)] , separate_axes = False)` 

> This could also be a good starting point on which to build a more generalized solution later?

I think the form of the API would need to be figured out first.  For instance, I'm not a fan of the Locator controlling if there are two axis levels or not.  The locator is really just there to give tick values. Rather it seems an `axis`-level method eg `xaxis.set_second_level_locator(BooLocator)`, or an `axes`-level method (ax.second_xaxis).

Overall, the problem of adding a second level of tick labels is largely an issue of how to place the tick labels in a way that looks good without manual tweaking.  It's a tricky problem in general.  There will also be the question of styling how the intervals in the second row are indicated.  

>  I think the readability/usability differences can be considered objectively.

I'm sure it is possible to come up with metrics for such things, and design studies to measure them.
